### PR TITLE
cli tests: Replace undo with op restore (pt 1)

### PR DESCRIPTION
**Pull request by ilyagr**
This is a test refactor. I intend to do this to all the tests, but I didn't have time today, so let's discuss the method and merge the portion I have done to avoid merge conflicts. 

This will be done to the remaining tests in a separate PR.

# Checklist

If applicable:
- [x] I have added tests to cover my changes

### CODE REVIEW

File: cli/tests/test_abandon_command.rs [hunk 1]
```diff
@@ -42,6 +42,7 @@ fn test_rebase_branch_with_merge() {
     create_commit(&test_env, &repo_path, "d", &["c"]);
     create_commit(&test_env, &repo_path, "e", &["a", "d"]);
     // Test the setup
+    test_env.advance_test_rng_seed_to_multiple_of(200_000);
```
[the remark this answers is not in the document]
Comment (ilyagr) [reply cli/tests/test_abandon_command.rs:46]: I was thinking of creating some helpers, and no longer using this function directly:

```
advance_rng_seed_to_next_section = advance_rng_seed_to_multiple_of(200_000)
advance_rng_seed_to_next_subsection = advance_rng_seed_to_multiple_of(10_000)
advance_rng_seed_to_next_subsubsection = advance_rng_seed_to_multiple_of(500)
advance_rng_seed_to_next_subsubsubsection = advance_rng_seed_to_multiple_of(25)
```

I was also thinking of replacing `test_rng_seed` with `change_id_rng`.

> I suppose one option is have let backup = test_env.backup() and test_env.restore(backup), which would copy the whole test environment directory aside.

I am not sure I follow -- this seems like a replacement for `op restore`, but it wouldn't help with the rng seed for change id generation, would it?

[the remark this answers is not in the document]
Comment (martinvonz) [reply cli/tests/test_abandon_command.rs:46]: > I am not sure I follow -- this seems like a replacement for `op restore`, but it wouldn't help with the rng seed for change id generation, would it?

Sorry, I wasn't clear. I was thinking that the backup would include a copy of the whole `TestEnvironment` struct, including the command number/seed.

[the remark this answers is not in the document]
Comment (ilyagr) [reply cli/tests/test_abandon_command.rs:46]: I'm not sure if this is a deal-breaker, but forking would still mean that if you make the test setup one command longer, all the change ids would change.

[the remark this answers is not in the document]
Comment (martinvonz) [reply cli/tests/test_abandon_command.rs:46]: Good point. Maybe `fork()` could bump the seed?

[the remark this answers is not in the document]
Comment (ilyagr) [reply cli/tests/test_abandon_command.rs:46]: I'm not sure that's better than `jj op restore` + functions to bump the seed, as we'd have now. I guess one advantage of forking is that it preserves the operations log. OTOH, I think using `jj op restore` in tests has a coolness factor.

[the remark this answers is not in the document]
Comment (martinvonz) [reply cli/tests/test_abandon_command.rs:46]: I don't mind using `jj op restore` in tests and I don't think we typically care to preserve the operation log. I just found don't like the arbitrary numbers. Even if we add helpers, we still need to remember which level "sub-heading" we need. Forking seems easier to understand. I might try it to see if it's easy to do or if it gets complicated

[the remark this answers is not in the document]
Comment (ilyagr) [reply cli/tests/test_abandon_command.rs:46]: I could add some guidance. The idea in my mind was that it's not very important. You can use `section` at first. If you are editing a test, and adding a `section` messes up more change ids than you'd like, use a `subsection`. And so on.

Most importantly, I would like to present it in such a way that people don't worry too much about it.

[the remark this answers is not in the document]
Comment (yuja) [reply cli/tests/test_abandon_command.rs:46]: > Before, it was easy to have a bug in tests where a command is accidentally
> non-mutating (or accidentally mutating), and so `undo` undid the wrong command.

For test stability, I think `test_env.fork()` is better option. It won't break if we messed up op undo/restore code.

[the remark this answers is not in the document]
Comment (ilyagr) [reply cli/tests/test_abandon_command.rs:46]: > It won't break if we messed up op undo/restore code.

I wouldn't mind tests breaking in this case -- this just means that our undo/restore code is tested more extensively.

In any case, there's no rush on this. I'm happy to have this cleanup wait until you bring forking to life. TBH, I still don't really follow how forking is better, but I'm biased and both of you seem to think it's a good idea.

File: cli/tests/common/mod.rs [hunk 2]
```diff
@@ -222,6 +222,13 @@ impl TestEnvironment {
         self.env_vars.insert(key.to_string(), val.to_string());
     }
 
+    pub fn current_operation_id(&self, repo_path: &Path) -> String {
+        let mut id = self.jj_cmd_success(repo_path, &["debug", "operation", "--display=id"]);
+        let len_trimmed = id.trim_end().len();
+        id.truncate(len_trimmed);
+        id
```
[the remark this answers is not in the document]
Comment (ilyagr) [reply cli/tests/common/mod.rs:229]: I fixed this and pulled it out to a separate https://github.com/martinvonz/jj/pull/2142.

**Pull request closed without merging**